integration-testing/client/CasperLabsClient/casperlabs_client/casperlabs_client.py

```diff
@@ -398,38 +442,31 @@ def deploy(
         if from_addr and len(from_addr) != 32:
             raise Exception(f"from_addr must be 32 bytes")
 
-        payment = payment or session
-
-        def hash(data: bytes) -> bytes:
-            h = blake2b(digest_size=32)
-            h.update(data)
-            return h.digest()
+        session_options = (session, session_hash, session_name, session_uref)
+        payment_options = (payment, payment_hash, payment_name, payment_uref)
 
-        def read_binary(file_name: str):
-            with open(file_name, "rb") as f:
-                return f.read()
+        # Compatibility mode, should be removed when payment is obligatory
+        if len(list(filter(None, payment_options))) == 0:
+            logging.info("No payment contract provided, using session as payment")
+            payment_options = session_options
 
-        def read_code(file_name: str, abi_encoded_args: bytes = None):
-            return consensus.Deploy.Code(
-                wasm=read_binary(file_name), abi_args=abi_encoded_args
+        if len(list(filter(None, session_options))) != 1:
+            raise TypeError(
+                "deploy: only one of session, session_hash, session_name, session_uref must be provided"
             )
 
-        def sign(data: bytes):
-            return private_key and consensus.Signature(
-                sig_algorithm="ed25519",
-                sig=ed25519.SigningKey(read_pem_key(private_key)).sign(data),
+        if len(list(filter(None, payment_options))) != 1:
+            raise TypeError(
+                "deploy: only one of payment, payment_hash, payment_name, payment_uref must be provided"
             )
 
-        def serialize(o) -> bytes:
-            return o.SerializeToString()
-
         # session_args must go to payment as well for now cause otherwise we'll get GASLIMIT error,
         # if payment is same as session:
         # https://github.com/CasperLabs/CasperLabs/blob/dev/casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala#L463
         body = consensus.Deploy.Body(
-            session=read_code(session, session_args),
-            payment=read_code(
-                payment, payment == session and session_args or payment_args
+            session=_encode_contract(session_options, session_args),
+            payment=_encode_contract(
+                payment_options, payment == session and session_args or payment_args
```

Comment:
```suggestion
                payment_options, payment_options == session_options and session_args or payment_args
```

I commented on this before but it seems like it's out of context: `session` and `payment` (i.e. the file paths) can both be `None` if they are given by other parameters, like `*_hash`. Still if those are not the same then using the same args is probably not the right thing to do.

Reply:
Corrected as suggested. 
This code is here only to support old tests that pass session wasm as payment, none of them passes the session or payment as a hash or name. Hopefully we can get rid of the no-payment mode and clean up the code here and everywhere.